crates/metal-llvm/src/lib.rs

```diff
@@ -1,3 +1,36 @@
 //! Metal library for compiling to LLVM IR using MIR.
 
-pub mod codegen;
+use metal_mir::parcel::Module;
+
+pub mod expr;
+pub mod primitives;
+pub mod stmt;
+pub mod ty;
+
+use llvm_sys::prelude::{LLVMBuilderRef, LLVMContextRef, LLVMModuleRef, LLVMTypeRef, LLVMValueRef};
+
+pub struct LLVMRefs {
+    ctx: LLVMContextRef,
+    builder: LLVMBuilderRef,
+    module: LLVMModuleRef,
+}
+
+pub trait CodeGenValue {
+    fn codegen_value(&self, llvm: &LLVMRefs, module: &Module) -> LLVMValueRef;
+}
+
+pub trait CodeGenType {
+    fn codegen_type(&self, llvm: &LLVMRefs, module: &Module) -> LLVMTypeRef;
+}
+
+pub fn get_types(
+    llvm: &LLVMRefs,
+    module: &Module,
+    types: Vec<metal_mir::types::Type>,
```

Comment:
Since `CodeGenType::codegen_type` only takes a reference to `self`, this should be changed to

```suggestion
    types: &[metal_mir::types::Type],
```

That way, the clone on L15 of `crates/metal-llvm/src/ty.rs` can be eliminated.